packages/adblocker/src/engine/engine.ts

```diff
@@ -229,6 +229,85 @@ export default class FilterEngine extends EventEmitter<EngineEventHandlers> {
     return engine as InstanceType<T>;
   }
 
+  public static fromEngines<T extends FilterEngine>(
+    this: new (...args: any[]) => T,
+    ...engines: T[]
+  ): T {
+    if (engines.length === 0) {
+      throw new Error(`merging engines requires at least two engines, use ".empty()" instead`);
+    }
+
+    const config = engines[0].config;
+    const lists = engines[0].lists;
+    const version = engines[0].version;
+
+    const networkFilters: Map<number, NetworkFilter> = new Map();
+    const cosmeticFilters: Map<number, CosmeticFilter> = new Map();
+    const preprocessors: Preprocessor[] = [];
+
+    const configKeysMustMatch: Exclude<keyof Config, 'serialize' | 'getSerializedSize'>[] = [
+      'debug',
```

Comment:
why only those 3?

Reply:
There's no problem to reuse objects if the given engines' configs are compatible with each other. Those three are the options impacting to the footprint size and behavior: e.g. `toString`.

The following options are safe to on and off on the runtime:

- For the options starts with `load`, we don't need to take care about them. Because it just means that the options are not loaded at the build time or parse time.
- `integrityCheck` is also not crucial to filter out. The only use of this value is in the deserialization function and we don't count it here.
- `guessRequestTypeFromUrl` is also not used in the adblocker library. It's used in the other projects like: `adblocker-webextension` and `adblocker-puppeteer`. This option can't affect to the engine behavior. Same goes for `enablePushInjectionsOnNavigationEvents`, `enableMutationObserver`, `enableInMemoryCache`, and `enableHtmlFiltering`.

Reply:
It does sound like at least `guessRequestTypeFromUrl`, `enablePushInjectionsOnNavigationEvents`, `enableMutationObserver`, `enableInMemoryCache`, `enableHtmlFiltering` do impact the behavior (maybe I misunderstood your message but I understood the other way around).

For the other ones, even if they don't impact the runtime behavior of the engine, it does beg the question as to which config value the resulting engine should have if they don't match and it's arbitrary to take the first one.

At least in the case of Ghostery I think the configs should always all match? So maybe we can start with that and add ignoring of some config values in the future if we need it.

Reply:
Updated in [`05c8b8f` (#4138)](https://github.com/ghostery/adblocker/pull/4138/commits/05c8b8f2b9977881be140127407d51316d2ab727).

Reply:
I think it'd make more sense to have an exclusion list here given that most fields are to be included. It'd be clearer and more future-proof to know which fields can be ignored from this check.

Reply:
Updated in [`461defd` (#4138)](https://github.com/ghostery/adblocker/pull/4138/commits/461defdd7e378379cdf0289788d6e2ce22ddb094)

Reply:
I think we can move the type and type guard to `config.ts` from `engine.ts` if we need it in another place in the future. I just inlined the type since it's only used in here.